bip-0324.mediawiki

```diff
@@ -0,0 +1,467 @@
+
+== Motivation ==
+
+The current peer-to-peer protocol is partially inefficient and in plaintext.
+
+With the current unencrypted message transport, BGP hijack, block delay attacks and message tampering are inexpensive and can be executed covertly (undetectable MITM)<ref>[https://btc-hijack.ethz.ch/files/btc_hijack.pdf Hijacking Bitcoin: Routing Attacks on Cryptocurrencies - M. Apostolaki, A. Zohar, L.Vanbever]</ref>.
+
+Adding opportunistic encryption introduces a high risk for attackers of being detected. Peer operators can compare encryption session IDs or use other form of authentication schemes <ref name="bip150">[https://github.com/bitcoin/bips/blob/master/bip-0150.mediawiki BIP150]</ref> to identify an attack.
+
+Each current version 1 Bitcoin peer-to-peer message uses a double-SHA256 checksum truncated to 4 bytes. Roughly the same amount of computation power would be required for encrypting and authenticating a peer-to-peer message with ChaCha20 & Poly1305.
+
+Additionally, this BIP describes a way to identify data which has been manipulated by peers (intercepting, then blocking or tampering with messages).
+
+Encrypting traffic between peers is already possible with VPN, tor, stunnel, curveCP or any other encryption mechanism on a deeper OSI level, however, most of those solutions require significant technical experience in setting up a secure channel and are therefore not widely deployed.
+
+== Specification ==
+
+<blockquote>
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in RFC 2119<ref>[https://tools.ietf.org/html/rfc2119 RFC 2119]</ref>.
+</blockquote>
+
+A peer that supports the message transport protocol as defined in this proposal MUST accept encryption requests from all peers.
+
+Both communication directions share the same shared-secret but have different symmetric cipher keys.
+
+The encryption handshake MUST happen before sending any other messages to the responding peer.
+
+If the responding peer closes the connection after sending the handshake request, the initiating peer MAY try to connect again with the v1 peer-to-peer transport protocol. Such reconnects allow an attacker to "downgrade" the encryption to plaintext communication and thus, accepting v1 connections MUST not be done when the Bitcoin peer-to-peer network has almost entirely embraced v2 communication.
+
+
+=== NODE_P2P_V2 ===
+
+Peers supporting the transport protocol after this proposal MUST signal <code>NODE_P2P_V2</code>
+<pre>
+NODE_P2P_V2 = (1 << 11)
+</pre>
+
+A peer usually learns an address along with the expected service flags which MAY be used to filter possible outbound peers.
+
+A peer signaling <code>NODE_P2P_V2</code> MUST accept encrypted communication specified in this proposal.
+
+Peers MAY only make outbound connections to peers supporting <code>NODE_P2P_V2</code>.
```

Comment:
Before you connect, you can't know whether the node supports V2. Why not say "maintain" instead of "make"?

Why is "peers" plural here, when it is singular above? An implementer of this BIP will write node software for a single peer, so I think singular works better. Better use plural for the other nodes.